File: experimentos/experimento/fl/estado.js
import { readFileSync, writeFileSync, existsSync } from "fs";
import { nombreAdministrador, claveAdministrador, administrador } from './constantes.js';

/** Modelo que representa el estado del sistema. */
export class Estado {
  usuarios;
  claves;
  saldos;
  sueldos;
  movimientos;
  saldo;

  constructor(usuarios, claves, saldos, sueldos, movimientos, saldo) {
    this.usuarios = usuarios;
    this.claves = claves;
    this.saldos = saldos;
    this.sueldos = sueldos;
    this.movimientos = movimientos;
    this.saldo = saldo;
  }

  /** Guarda la instancia actual del modelo de Estado en el archivo `estado.json`. */
  guardar() {
    let json = {};
    json["usuarios"] = this.usuarios;
    json["claves"] = this.claves;
    json["saldos"] = this.saldos;
    json["sueldos"] = this.sueldos;
    json["movimientos"] = this.movimientos;
    json["saldo"] = this.saldo;
    writeFileSync("estado.json", JSON.stringify(json), { encoding: "utf8" });
  }

  /** Retorna una instancia del modelo de estado con los valores guardados en `estado.json`. */
  static cargar() {
    let contenido = "";
    if (existsSync("estado.json")) {
      contenido = readFileSync("estado.json", { encoding: "utf8" }).toString();
    }

    if (contenido === "") {
      return Estado.#inicial();
    } else {
      let json = JSON.parse(contenido);
      let estado = new Estado();

      estado.usuarios = json["usuarios"];
      estado.claves = json["claves"];
      estado.saldos = json["saldos"];
      estado.sueldos = json["sueldos"];
      estado.movimientos = json["movimientos"];
      estado.saldo = json["saldo"];
      return estado;
    }
  }
  /** Retorna una instancia del modelo de estado con sus valores iniciales. */
  static #inicial() {
    let estado = new Estado();

    estado.usuarios = [];
    estado.claves = [];
    estado.saldos = [];
    estado.sueldos = [];
    estado.movimientos = [];
    estado.saldo = 0;
    return estado;
  }
}

File: experimentos/experimento/gt/Estado.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.Json;

namespace Solucion;


/// <summary>
/// Modelo que representa el estado del sistema.
/// </summary>
public class Estado
{
    // Completar con las propiedades del estado.
    public Dictionary<int, string> usuarios { get; set; }
    public Dictionary<int, string> claves { get; set; }
    public Dictionary<int, int> saldos { get; set; }
    public Dictionary<int, int> sueldos { get; set; }
    public Dictionary<int, Dictionary<DateTime, OPERACION>> movimientos { get; set; }
    public int saldo { get; set; }

    /// <summary>
    /// Guarda la instancia actual del modelo de Estado en el archivo `estado.json`.
    /// </summary>
    public void Guardar()
    {
        string ruta = "C:\\Users\\fs80235\\Desktop\\atm\\experimentos\\experimento\\Galli-Tomas\\estado.json";
        if (ruta.StartsWith("/"))
        {
            ruta = "." + ruta;
        }
        string contenido = JsonSerializer.Serialize(this);
        File.WriteAllText(ruta, contenido, Encoding.UTF8);
    }

    /// <summary>
    /// Retorna una instancia del modelo de estado con los valores guardados en `estado.json`.
    /// </summary>
    public static Estado Cargar()
    {
        string ruta = "C:\\Users\\fs80235\\Desktop\\atm\\experimentos\\experimento\\Galli-Tomas\\estado.json";
        string contenido = "";
        if (File.Exists(ruta))
        {
            contenido = File.ReadAllText(ruta, Encoding.UTF8);
        }

        if (contenido == "")
        {
            return Inicial();
        }
        else
        {
            return JsonSerializer.Deserialize<Estado>(contenido);
        }
    }

    /// <summary>
    /// Retorna una instancia del modelo de estado con sus valores iniciales.
    /// </summary>
    private static Estado Inicial()
    {
        var estado = new Estado();

        estado.usuarios = new Dictionary<int, string>
        {
            { Globales.administrador, Globales.nombre_administrador }
        };
        estado.claves = new Dictionary<int, string>
        {
            { Globales.administrador, Globales.clave_administrador }
        };
        estado.saldos = new Dictionary<int, int>();
        estado.sueldos = new Dictionary<int, int>();
        estado.movimientos = new Dictionary<int, Dictionary<DateTime, OPERACION>>();
        estado.saldo = 0;

        return estado;
    }

    public static class Globales
    {
        public readonly static int CANT_MAX_USUARIOS = 5;
        public readonly static int LONG_MIN_CLAVE = 8;
        public readonly static int administrador = 12345678;
        public readonly static string nombre_administrador = "admin";
        public readonly static string clave_administrador = "admin123";
        public static DateTime ahora => DateTime.Now;
    }
}

/* --- Ejemplo de uso ---
Estado e = Estado.Cargar();
...
e.Guardar();
*/
